File: src/infrastructure/terminal/terminal-ports.ts
export interface TerminalOutput {
  write(message: string): void
  writeError(message: string): void
}

export interface ProgressReporter {
  start(message: string): void
  message(message: string): void
  stop(message: string): void
  error(message: string): void
}

export interface PromptPort {
  text(options: {
    message: string
    initialValue?: string
    defaultValue?: string
    placeholder?: string
    validate?: (value: string | undefined) => string | undefined
  }): Promise<string | undefined>
  password(options: {
    message: string
    validate?: (value: string | undefined) => string | undefined
  }): Promise<string | undefined>
  select(options: {
    message: string
    options: Array<{ value: string; label: string; hint?: string; disabled?: boolean }>
    initialValue?: string
  }): Promise<string | undefined>
  confirm(options: { message: string; initialValue?: boolean }): Promise<boolean | undefined>
}

File: lib/src/application/terminal/terminal_ports.dart
typedef PromptValidator = String? Function(String value);

final class PromptOption {
  const PromptOption({required this.value, required this.label, this.hint});

  final String value;
  final String label;
  final String? hint;
}

abstract interface class PromptPort {
  String? text({
    required String message,
    String? initialValue,
    String? placeholder,
    PromptValidator? validate,
  });

  String? password({required String message});

  String? select({
    required String message,
    required List<PromptOption> options,
    String? initialValue,
  });
}

abstract interface class TerminalOutput {
  void heading(String title, {String? detail});

  void card(String title, String content);

  void write(String message);

  void writeError(String message);

  void info(String message);

  void success(String message);

  void warning(String message);

  void detail(String message);
}

abstract interface class ProgressReporter {
  void start(String message);

  void message(String message);

  void stop(String message);

  void error(String message);
}
